File: dotnet/tests/Vulgarity.Tests/TestData.cs
using System;
using System.IO;
using System.Text.Json;

namespace Vulgarity.Tests
{
    /// <summary>Finds the shared data directory that both ports read.</summary>
    internal static class TestData
    {
        private static readonly Lazy<string> Root = new Lazy<string>(Locate);

        public static string Directory
        {
            get { return Root.Value; }
        }

        private static string Locate()
        {
            DirectoryInfo dir = new DirectoryInfo(AppContext.BaseDirectory);
            while (dir != null)
            {
                string candidate = Path.Combine(dir.FullName, "data");
                if (File.Exists(Path.Combine(candidate, "seed.json")))
                {
                    return candidate;
                }

                dir = dir.Parent;
            }

            throw new InvalidOperationException("Cannot find the shared data directory.");
        }

        public static string Read(string name)
        {
            return File.ReadAllText(Path.Combine(Directory, name));
        }

        public static byte[] ReadBytes(string name)
        {
            return File.ReadAllBytes(Path.Combine(Directory, name));
        }

        public static JsonDocument ReadJson(string name)
        {
            return JsonDocument.Parse(Read(name));
        }

        /// <summary>The number of terms the English seed holds.</summary>
        /// <remarks>
        /// A filter built from seed.json must report exactly this many. Read it
        /// here rather than writing the number into a test: the seed is
        /// generated, it grows, and a literal only records what it happened to
        /// be on the day.
        /// </remarks>
        public static int SeedEntryCount()
        {
            using (JsonDocument doc = ReadJson("seed.json"))
            {
                return doc.RootElement.GetProperty("entries").GetArrayLength();
            }
        }

        /// <summary>The system word list, or null when this machine has none.</summary>
        public static string[] EnglishWords()
        {
            const string path = "/usr/share/dict/words";
            return File.Exists(path) ? File.ReadAllLines(path) : null;
        }
    }
}

File: dart/test/test_data.dart
import 'dart:convert';
import 'dart:io';

/// Finds the shared data directory that both ports read.
final Directory dataDirectory = _locate();

Directory _locate() {
  Directory dir = Directory.current;
  while (true) {
    final Directory candidate = Directory('${dir.path}/data');
    if (File('${candidate.path}/seed.json').existsSync()) {
      return candidate;
    }
    final Directory parent = dir.parent;
    if (parent.path == dir.path) {
      throw StateError('Cannot find the shared data directory.');
    }
    dir = parent;
  }
}

String readData(String name) =>
    File('${dataDirectory.path}/$name').readAsStringSync();

List<int> readDataBytes(String name) =>
    File('${dataDirectory.path}/$name').readAsBytesSync();

Map<String, dynamic> readJson(String name) =>
    jsonDecode(readData(name)) as Map<String, dynamic>;

Map<String, dynamic> readFixture(String name) =>
    jsonDecode(File('${dataDirectory.path}/testdata/$name').readAsStringSync())
        as Map<String, dynamic>;

/// The system word list, or null when this machine has none.
List<String>? englishWords() {
  final File file = File('/usr/share/dict/words');
  return file.existsSync() ? file.readAsLinesSync() : null;
}
